DOC: add AI policy (#24583)
scipy/scipy: doc/source/dev/conduct/ai_policy.rst

AI Policy
=========

"AI" herein refers to generative AI tools like large language models (LLMs)
that can generate, edit, and review software code, create and manipulate
images, or generate human-like communication.

Responsibility
==============

You are responsible for any code you submit to SciPy's repositories, regardless
of whether it was manually written or generated by AI. You must understand and be able
to explain the code you submit as well as the existing related code. It is not
acceptable to submit a patch that you cannot understand and explain yourself.
In explaining your contribution, do not use AI to automatically generate
descriptions, as AI rarely communicates such information correctly and

> concisely.

Reviewer: I don't think we even need to justify the requirement.

> concisely.

Author: Justification removed

Disclosure
==========

If you make use of AI to assist in the development of your patch,
you must disclose how it was used and what code in the patch is AI generated.
We will reject any pull request that does not include the disclosure.

Code Quality
============

Code generated by AI is very often of low quality. Contributors are expected to
submit code that meets SciPy's standards. We will reject pull requests that we
deem being "`AI slop`_". Do not waste developers' time by submitting code that
is fully or mostly generated by AI.

.. _AI slop: https://en.wikipedia.org/wiki/AI_slop

Copyright
=========

All code in SciPy is released under the BSD 3-clause copyright license.
Contributors to SciPy license their code under the same license when it is
included into SciPy's version control repository. That means contributors must
own the copyright of any code submitted to SciPy or must include the BSD
3-clause compatible open source license(s) associated with the submitted code
in the patch. Code generated by AI may infringe on copyright and it is the
submitter's responsibility to not infringe. We will reject any pull requests,

> AI generated or not, where the copyright is in question.

Reviewer 2: What would you think about a lighter phrasing:

>We will reject any pull requests where the copyright infringes upon the license of a work inside the AI's training set.

This narrows it from "where the copyright is in question" to "where the copyright infringes." As you say, the copyright may always be *questionable*. However, I think it is useful to clarify to a contributor that we care about copyright before they embark on some huge rewrite that will be rejected out of hand for being too hard to verify non-infringement.

(Why did I add the phrase "upon the license of a work inside the AI's training set?" Sometimes people come to the project with the idea that they can disclaim ownership of the AI's output, and that therefore solves our concern about copyright. That's not our concern, our concern is about training materials.)

> AI generated or not, where the copyright is in question.

Reviewer 3: right, I'm just still uncomfortable with the language of

> We _will_ reject any pull requests where ...

due to the apparent vagueness or indeterminacy. Perhaps "we _reserve the right to_ reject ..." would be better? I'm still not sure how to follow it, though.

> AI generated or not, where the copyright is in question.

Reviewer 3: > I think this dispute might be based upon differing views for the intended audience of this documentation. I view the intended audience as a new contributor, who has maybe participated in one or two issues, but has not read most issues or most PRs. I don't see this policy as being a directive to reviewers.

Oh no, I totally agree with you there. My point is precisely that the language "we will ...", where `"we" == maintainers`, _unintentionally_ dispenses expectation on maintainers (rather than contributors).

---

> I do think the existing language conveys that. 'Questionable' gives reviewers some level of flexibility - if they merged it, evidently they didn't think it was questionable.

I don't think this squares up with my personal position at all. I think that I simultaneously hold:
1) since any PR being presented to me could plausibly contain LLM output, and along the lines of https://mail.python.org/archives/list/numpy-discussion@python.org/message/7L3WL63ACL6OJPP3ZQ6AA4D6CZFZSNAC/, I find it plausible that almost every PR I review is _questionable_ (for all I know, and to a significantly greater degree than pre-LLMs) w.r.t. copyright
2) I am willing to merge (i.e. not morally/legally/... barred in my estimation) most PRs I review (eventually)

I guess my point is that:
- (a) In the venn-diagram of 'questionable w.r.t. copyright' and 'happy to merge' (per typical maintainer intuition), the intersection is plausibly increasingly large (per (1))
- (b) The actual intention behind this section of text is to target the set of PRs which fall _inside_ the 'questionable w.r.t copyright' set and _outside_ the 'happy to merge' set
- (c) The current text appears to target the entire 'questionable w.r.t. copyright' set rather than the set identified in (b)

I think you agree with (b)? Maybe you disagree with (a), but I think it is at least plausible. Maybe (c) is just me having an unusual reading, but maybe not?

> AI generated or not, where the copyright is in question.

Reviewer 4: (sorry this is a very brief comment, I have no time to review the whole PR in detail today, but this stood out to me)

> I view the intended audience as a new contributor,

This is written as a SciPy project _policy_, so by definition it applies to everyone. Hence it needs to be phrased and read as such. If you want some more handwaving or different guidance specific to people not yet familiar with the project, let's do that separately in a "getting started" part of the contributor guide, referencing this policy.

> AI generated or not, where the copyright is in question.

Reviewer 2: > I think you agree with (b)? Maybe you disagree with (a), but I think it is at least plausible. Maybe (c) is just me having an unusual reading, but maybe not?

I agree with a, b, and c. Not sure how to improve that. Will think on it.

> This is written as a SciPy project _policy_, so by definition it applies to everyone.

Understood.

Communication
=============

When interacting in communication among developers (forum, discussions,
issues, pull requests, etc.) do not use AI to speak for you, other than for
translation or grammar editing. If the developers want to chat with a chatbot,
they can do so themselves. Human-to-human communication is essential for an
open source community to thrive.

AI Agents
=========

> The use of AI agents who write code and then submit pull requests is not permitted.

Author: ```suggestion
The use of AI agents that writes code and then submit a pull request is not permitted.
```

> As per the responsibilities section, a human contributor must be responsibility for

Reviewer 5: ```suggestion
As per the responsibilities section, a human contributor must be responsible for
```

any code submitted.

New Contributors
================

A high quality patch to SciPy that follows our development practices and
includes human communication to get it merged may be assisted by AI, but in our
experience, this requires a seasoned developer to properly execute. If you are
a new contributor to SciPy, we will most likely close any issues, discussions,
or pull requests that are substantially generated by AI.

Other Resources
===============
While these do not formally form part of SciPy's AI policy, the following resources
may be helpful in understanding some pitfalls associated with using AI to contribute to
SciPy:

- https://llvm.org/docs/AIToolPolicy.html
- https://blog.scientific-python.org/scientific-python/community-considerations-around-ai/

Acknowledgements
================
We thank the SymPy developers for their AI policy, which this document is largely based

> upon.

Reviewer 3: I suppose we could add a hyperlink to their policy?

```suggestion
We thank the SymPy developers for their AI policy, upon which this document is largely
based.
```

AI Policy
=========

"AI" herein refers to generative AI tools like large language models (LLMs)
that can generate, edit, and review software code, create and manipulate
images, or generate human-like communication.

Responsibility
--------------

You are responsible for any code you submit to SciPy's repositories, regardless
of whether it was manually written or generated by AI. You must understand and be able
to explain the code you submit as well as the existing related code. It is not
acceptable to submit a patch that you cannot understand and explain yourself.
In explaining your contribution, do not use AI to automatically generate
descriptions.

Disclosure
----------

If you make use of AI to assist in the development of your pull request,
you must disclose disclose the tool(s) used, how they were used, and specify what code
or text is AI generated. We will reject any pull request that does not include the
disclosure.

Code Quality
------------

Code generated by AI can be of low quality. Contributors are expected to
submit code that meets SciPy's standards. We will reject pull requests that we
deem being "`AI slop`_". Do not waste developers' time by submitting code that

> is fully or mostly generated by AI.

Reviewer 4: ```suggestion
is fully or mostly generated by AI, and doesn't meet our standards.
```

Otherwise this won't age will, or even won't work well today. If say boilerplate code is fully AI-generated and then verified for quality by a contributor before submission, that's fine. I actually expect that _if_ I'd use AI for codegen, that's how I'd start - simple stuff that's easy to automate and check.

.. _AI slop: https://en.wikipedia.org/wiki/AI_slop

Copyright
---------

All code in SciPy is released under the BSD 3-clause copyright license.
Contributors to SciPy license their code under the same license when it is
included into SciPy's version control repository. That means contributors must
own the copyright of any code submitted to SciPy or must include the BSD
3-clause compatible open source license(s) associated with the submitted code
in the patch. Code generated by AI may infringe on copyright and it is the
submitter's responsibility to not infringe. We reserve the right to reject any pull
requests, AI generated or not, where the copyright is in question.

Communication
-------------

When interacting in communication among developers (forum, discussions,
issues, pull requests, etc.) do not use AI to speak for you, other than for

> translation or grammar editing. If the developers want to chat with a chatbot,

Reviewer 6: simpler wording maybe? 

```suggestion
When interacting with developers (forum, discussions,
issues, pull requests, etc.) do not use AI to speak for you, except for
translation or grammar editing. If the developers want to chat with a chatbot,
```
